cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java

```diff
@@ -48,56 +48,37 @@ private ConfigProducerUtil() {
             return null;
         }
 
+        String defaultValue = getDefaultValue(injectionPoint);
+        String resolvedValue = resolveValue(name, config, defaultValue);
         if (hasCollection(injectionPoint.getType())) {
-            return convertValues(name, injectionPoint.getType(), getDefaultValue(injectionPoint), config);
+            return convertValues(name, injectionPoint.getType(), resolvedValue, defaultValue, config);
         }
 
-        return convertValue(name, resolveConverter(injectionPoint, config), getDefaultValue(injectionPoint), config);
+        return SmallRyeConfig.convertValue(name, resolvedValue, resolveConverter(injectionPoint, config));
     }
 
     public static <T> T getValue(String name, Type type, String defaultValue, Config config) {
         if (name == null) {
             return null;
         }
-        return convertValue(name, resolveConverter(type, config), defaultValue, config);
+        String resolvedValue = resolveValue(name, config, defaultValue);
+        return SmallRyeConfig.convertValue(name, resolvedValue, resolveConverter(type, config));
     }
 
-    public static <T> T convertValue(String name, Converter<T> converter, String defaultValue, Config config) {
-        String rawValue = getRawValue(name, config);
-        if (rawValue == null) {
-            rawValue = defaultValue;
-        }
-
-        T converted;
-        if (rawValue == null) {
-            // convert an empty value
-            try {
-                converted = converter.convert("");
-            } catch (IllegalArgumentException ignored) {
-                throw InjectionMessages.msg.propertyNotFound(name);
-            }
-        } else {
-            converted = converter.convert(rawValue);
-        }
-        if (converted == null) {
-            throw InjectionMessages.msg.propertyNotFound(name);
-        }
-        return converted;
-    }
-
-    public static <T> T convertValues(String name, Type type, String defaultValue, Config config) {
+    public static <T> T convertValues(String name, Type type, String resolvedValue, String defaultValue, Config config) {
         String rawValue = getRawValue(name, config);
```

Comment:
I see. I think this will need a little shuffle round of methods in `ConfigProducerUtil` due to this edge case in indexedProperties: https://github.com/smallrye/smallrye-config/blob/3e04ed5be90e31fab3a30342d8118b24f6640711/cdi/src/test/java/io/smallrye/config/inject/IndexedPropertiesInjectionTest.java#L96-L98